WordPress/gutenberg
# Fix broken undo history stack for Pattern Overrides

## What?

The undo history is a bit broken for Pattern Overrides. This PR tries to fix it.

## Why?

For better editing experience.

## How?

This PR removes the `syncDerivedBlockAttributes` action introduced in the original PR (#56235) an instead introduce a new private higher-order action `syncDerivedUpdates`.

The signature of `syncDerivedUpdates` looks like:
```ts
syncDerivedUpdates( callback: () => void ): void;
```

Any updates inside the callback should not push additional undo history stacks nor should be persistent. This serves as a escape hatch to make updates to the block editor completely invisible to the core's undo manager. This means we can call `setAttributes` to set the `overrides` of the pattern block without worrying that it might break the history stack.

Example usage:
```js
const { syncDerivedUpdates } = unlock( registry.dispatch( blockEditorStore ) );

syncDerivedUpdates( () => {
  setAttributes({ overrides });
} );
```

The goal is to make something like this stable enough so that third-party developers can leverage it too. We probably can use it solve some of the longstanding undo bugs too (https://github.com/WordPress/gutenberg/issues/8119#issuecomment-1185238067, https://github.com/WordPress/gutenberg/issues/41031#issuecomment-1188544191).

> [!NOTE]
> This API currently doesn't work 100% of the time for other blocks so it should remain private for now. The top priority of this PR is to fix the existing issue for Pattern Overrides. We can work on finalizing the API across different use cases in the future.

## Testing Instructions




1. Enable the "Partial syncing patterns gutenberg experiment.
2. Go to Site Editor -> Patterns -> Create a synced pattern.
3. Add a paragraph and go to the block inspector sidebar -> Advanced -> Check the "Synced attributes" checkbox.
4. Save the pattern.
5. Go to edit a post.
6. Insert the pattern.
7. Make some changes to the enabled paragraph and test the undo history.

## Screenshots or screencast 


https://github.com/WordPress/gutenberg/assets/7753001/0f95849b-0076-48ec-9687-98fb5bc46a1d
